rfq/oracle.go

```diff
@@ -105,55 +115,183 @@ type PriceOracle interface {
-//	transportCredentials := credentials.NewTLS(&tlsConfig)
-//	opts = append(opts, grpc.WithTransportCredentials(transportCredentials))
-//
-//	return opts, nil
-// }
-//
-// // NewRpcPriceOracle creates a new RPC price oracle handle given the address
-// // of the price oracle RPC server.
-// func NewRpcPriceOracle(addr url.URL) (*RpcPriceOracle, error) {
-//	//// Connect to the RPC server.
-//	//dialOpts, err := serverDialOpts()
-//	//if err != nil {
-//	//	return nil, err
-//	//}
-//	//
-//	//serverAddr := fmt.Sprintf("%s:%s", addr.Hostname(), addr.Port())
-//	//conn, err := grpc.Dial(serverAddr, dialOpts...)
-//	//if err != nil {
-//	//	return nil, err
-//	//}
-//
-//	return &RpcPriceOracle{}, nil
-// }
-//
-// // QueryAskingPrice returns the asking price for the given asset amount.
-// func (r *RpcPriceOracle) QueryAskingPrice(ctx context.Context,
-//	assetId *asset.ID, assetGroupKey *btcec.PublicKey, assetAmount uint64,
-//	bidPrice *lnwire.MilliSatoshi) (*OracleAskResponse, error) {
-//
-//	//// Call the external oracle service to get the exchange rate.
-//	//conn := getClientConn(ctx, false)
-//
-//	return nil, nil
-// }
-//
-// // Ensure that RpcPriceOracle implements the PriceOracle interface.
-// var _ PriceOracle = (*RpcPriceOracle)(nil)
+// RpcPriceOracle is a price oracle that uses an external RPC server to get
+// exchange rate information.
+type RpcPriceOracle struct {
+	// client is the RPC client that this instance will use to interact with
+	// the price oracle RPC server.
+	client oraclerpc.PriceOracleClient
+
+	// rawConn is the raw connection to the remote gRPC service.
+	rawConn *grpc.ClientConn
+}
+
+// serverDialOpts returns the set of server options needed to connect to the
+// price oracle RPC server using a TLS connection.
+func serverDialOpts() ([]grpc.DialOption, error) {
+	var opts []grpc.DialOption
+
+	// Skip TLS certificate verification.
+	tlsConfig := tls.Config{InsecureSkipVerify: true}
+	transportCredentials := credentials.NewTLS(&tlsConfig)
+	opts = append(opts, grpc.WithTransportCredentials(transportCredentials))
+
+	return opts, nil
+}
+
+// NewRpcPriceOracle creates a new RPC price oracle handle given the address
+// of the price oracle RPC server.
+func NewRpcPriceOracle(addrStr string) (*RpcPriceOracle, error) {
+	addr, err := ParsePriceOracleAddress(addrStr)
+	if err != nil {
+		return nil, err
+	}
+
+	// Connect to the RPC server.
+	dialOpts, err := serverDialOpts()
+	if err != nil {
+		return nil, err
+	}
+
+	serverAddr := fmt.Sprintf("%s:%s", addr.Hostname(), addr.Port())
+	conn, err := grpc.Dial(serverAddr, dialOpts...)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create a new price oracle client from the active connection.
+	client := oraclerpc.NewPriceOracleClient(conn)
+
+	return &RpcPriceOracle{
+		client:  client,
+		rawConn: conn,
+	}, nil
+}
+
+// QueryAskPrice returns the ask price for the given asset amount.
+func (r *RpcPriceOracle) QueryAskPrice(ctx context.Context,
```

Comment:
We can move this into the main package along with the init information above, then the RFQ system accepts a fully initialized version of the service.

Reply:
I'm going to work on a test for this. If there's time before the release I'll take a look at the change you suggest. Let me know if you'd like me to prioritise differently.